content/articles/get-started-with-a-lexical-analyzer/index.md

```diff
@@ -0,0 +1,534 @@
+
+The output on the terminal is as shown below:
+
+![Terminal code run output](/engineering-education/get-started-with-a-lexical-analyzer/terminal-output.png "Terminal code run output")
+
+> YACC comes with predefined language reserved words, also known as keywords. Among these terms are "_yytext_," "_yylex_," and "_yywrap_", to name a few.
+> As one can notice, they start with `yy`. This step is because YACC uses that prefix before keywords to recognize them by default. One can also set the program to use another prefix of two characters. The characters can be lowercase, one uppercase, and the other lowercase.
+> Both can not be uppercase since all uppercase combinations have been used for other keywords.
+
+`yylex()` is a function that passes tokens found to `yyparse()` each time it is called. It returns 0 or a negative value whenever an EOF (End of File) error is returned. It utilizes a routine called `yygetc()` to fetch these values.
+`yyparse()` calls upon `yylex()` to obtain any tokens passed to the program. It then matches it against the patterns used for matches to determine the kind of action to be performed.
+`yywrap()` is used to show the end of the execution of the code. It returns a value of 1 if the program's run is over and a value of 0 if it is ongoing.
+
+Yacc also has some predefined variables. These variables include the following:
+- **yytext**: It defines the current recognizable input token by the lexical scanner.
+- **yylen**: It is a variable that holds the length of the input text.
+- **yyout**: It determines the output stream for the output macro that processes input that does not match any rules set.
+- **yyin**: It determines the input stream for the input functions and `yylex()`.
+- **yylineno**: It specifies the current input line number, which is kept up to date by the input and the `yycomment`.
+
+> `yyin` and `yyout` can be interchanged. This interchanging can be done by assignment.
+
+### A Lex analyzer that converts instances of certain characters into a particular pattern
+
+Let us create a lex analyzer that changes 'abc' occurrences in the input to 'ABC'.
+
+- Accomplish this by creating a file named `strings.l`.
+
+- In it, add the following in the definitions section:
+
+```c
+%{
+    /* Find instances of 'abc' to 'ABC' */
+
+#include<stdio.h>
+
+#include<string.h>
+
+int i;
+
+%}
+
+%%
+```
+> As for comments in a lex file, make sure to begin with double-spacing first to avoid errors.
+
+- Add rules to check on words only entered (with either upper or lowercase) containing the characters 'abc' following each other.
+  If yes, then replace the specified characters only.
+
+```c
+[a-z A-Z]* {
+
+for(i=0;i<=yyleng;i++)
+
+{
+
+if((yytext[i]=='a')&&(yytext[i+1]=='b')&&(yytext[i+2]=='c'))
+
+{
+
+yytext[i]='A';
+
+yytext[i+1]='B';
+
+yytext[i+2]='C';
+
+}
+
+}
+
+printf("%s",yytext);
+
+}
+
+.* {ECHO;}
+
+\n {printf("%s",yytext);}
+
```

Comment:
Format these

Reply:
It's quite hard; when you add indentation on lexical files, the whole app breaks. They are very sensitive to structure and indentation